Sources/APIServer/Networks/NetworksService.swift

```diff
@@ -161,28 +172,61 @@ actor NetworksService {
             throw ContainerizationError(.invalidState, message: "cannot delete subnet \(id) in state \(networkState.state)")
         }
 
-        let client = NetworkClient(id: id)
-        guard try await client.disableAllocator() else {
-            throw ContainerizationError(.invalidState, message: "cannot delete subnet \(id) with containers attached")
-        }
+        // prevent container operations while we atomically check and delete
+        try await containersService.withContainerList { containers in
+            // find all containers that refer to the network
+            var referringContainers = Set<String>()
+            for container in containers {
+                for containerNetworkId in container.configuration.networks {
+                    if containerNetworkId == id {
+                        referringContainers.insert(container.configuration.id)
+                        break
+                    }
+                }
+            }
 
-        defer { networkStates.removeValue(forKey: id) }
-        do {
-            try pluginLoader.deregisterWithLaunchd(plugin: networkPlugin, instanceId: id)
-        } catch {
-            log.error(
-                "failed to deregister network service after failed creation",
-                metadata: [
-                    "id": "\(id)",
-                    "error": "\(error.localizedDescription)",
-                ])
-        }
+            // bail if any referring containers
+            guard referringContainers.isEmpty else {
+                throw ContainerizationError(
+                    .invalidState,
+                    message: "cannot delete subnet \(id) with referring containers: \(referringContainers.joined(separator: ", "))"
+                )
+            }
 
-        do {
-            try await store.delete(id)
-        } catch {
-            throw ContainerizationError(.notFound, message: error.localizedDescription)
+            // disable the allocator so nothing else can attach
+            // TODO: remove this from the network helper later, not necesssary now that withContainerList is here
+            let client = NetworkClient(id: id)
+            guard try await client.disableAllocator() else {
+                throw ContainerizationError(.invalidState, message: "cannot delete subnet \(id) with active containers")
```

Comment:
nit: would this be a better error message?
```suggestion
                throw ContainerizationError(.invalidState, message: "cannot delete subnet \(id) - network allocator could not be disabled")
```